vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_util.cc

```diff
@@ -0,0 +1,269 @@
+
+#include <utility>
+
+#include "base/base64.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/global_constants.h"
+#include "bat/ledger/internal/common/random_util.h"
+#include "bat/ledger/internal/gemini/gemini_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "crypto/random.h"
+
+namespace ledger {
+namespace gemini {
+
+std::string GetClientId() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_WALLET_CLIENT_ID
+             : GEMINI_WALLET_STAGING_CLIENT_ID;
+}
+
+std::string GetClientSecret() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_WALLET_CLIENT_SECRET
+             : GEMINI_WALLET_STAGING_CLIENT_SECRET;
+}
+
+std::string GetUrl() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? GEMINI_OAUTH_URL
+             : GEMINI_OAUTH_STAGING_URL;
+}
+
+std::string GetFeeAddress() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? kFeeAddressProduction
+             : kFeeAddressStaging;
+}
+
+std::string GetACAddress() {
+  return ledger::_environment == type::Environment::PRODUCTION
+             ? kACAddressProduction
+             : kACAddressStaging;
+}
+
+std::string GetAuthorizeUrl(const std::string& state) {
+  const std::string id = GetClientId();
+  const std::string url = GetUrl();
+
+  return base::StringPrintf(
+      "%s/auth"
+      "?client_id=%s"
+      "&scope="
+      "balances:read,"
+      "history:read,"
+      "crypto:send,"
+      "account:read,"
+      "payments:create,"
+      "&redirect_uri=rewards://gemini/authorization"
+      "&state=%s"
+      "&response_type=code",
+      url.c_str(), id.c_str(), state.c_str());
+}
+
+std::string GetVerifyUrl() {
+  const std::string url = GetUrl();
+
+  return base::StringPrintf("%s/register/verify", url.c_str());
+}
+
+std::string GetAddUrl() {
+  return GetAccountUrl();
+}
+
+std::string GetWithdrawUrl() {
+  return GetAccountUrl();
+}
+
+type::ExternalWalletPtr GetWallet(LedgerImpl* ledger) {
+  DCHECK(ledger);
+  const std::string wallet_string =
+      ledger->ledger_client()->GetEncryptedStringState(state::kWalletGemini);
+
+  if (wallet_string.empty()) {
+    return nullptr;
+  }
+
+  base::Optional<base::Value> value = base::JSONReader::Read(wallet_string);
+  if (!value || !value->is_dict()) {
+    BLOG(0, "Parsing of Gemini wallet failed");
+    return nullptr;
+  }
+
+  base::DictionaryValue* dictionary = nullptr;
+  if (!value->GetAsDictionary(&dictionary)) {
+    BLOG(0, "Parsing of Gemini wallet failed");
+    return nullptr;
+  }
+
+  auto wallet = ledger::type::ExternalWallet::New();
+  wallet->type = constant::kWalletGemini;
+
+  auto* token = dictionary->FindStringKey("token");
```

Comment:
none of these are required fields?

Reply:
right! I have filed a follow-up issue for `one_time_string` - https://github.com/brave/brave-browser/issues/16766

Reply:
omg I didn't even realize we have `GetWallet` for Bitflyer and Uphold that are all identical except for wallet_type. This is not a follow-up, we're not going to copy all of this json parsing code a 3rd time.

Reply:
anyway I think you're missing my point here which is that some of these fields must be required or this object would be meaningless